### PR TITLE
sts-disconnect

### PR DESCRIPTION
**Pull request by yunjiang29**
https://issues.redhat.com/browse/OCPQE-12332 Need a CI profile to specify AWS Service Endpoint in STS env

### CODE REVIEW

**Comment by jianlinliu** [comment]
I saw we already have `cucushift-installer-rehearse-aws-ipi-proxy-cco-manual-security-token-service-provision`, does that means, sts + behind proxy would not hit the issues?

**Comment by yunjiang29** [comment]
Proxy cluster allows both global and regional STS endpoint, but disconnected cluster only allow regional STS endpoint, some operators always access global endpoint (also in disconnected cluster) that caused the problem.

File: ci-operator/step-registry/ipi/conf/aws/custom-endpoints/ipi-conf-aws-custom-endpoints-commands.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+REGION="${LEASED_RESOURCE}"
+CONFIG="${SHARED_DIR}/install-config.yaml"
+
+function patch_endpoint()
+{
+  local service_name=$1
+  local service_endpoint=$2
+  local config_patch="${SHARED_DIR}/install-config-${service_name}.yaml.patch"
+  if [ "$service_endpoint" == "DEFAULT_ENDPOINT" ]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/aws/custom-endpoints/ipi-conf-aws-custom-endpoints-commands.sh:11]: How about when `service_endpoint` is "", we treat that as users want to default endpoint, once that, we can drop `DEFAULT_ENDPOINT` here.

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/custom-endpoints/ipi-conf-aws-custom-endpoints-commands.sh:11]: That means once users called `ipi-conf-aws-custom-endpoints`, but did not set a custom `SERVICE_ENDPOINT_*`, the script will set the endpoints to `https://${service_name}.${REGION}.amazonaws.com`

**Comment by jianlinliu** [comment]
> Proxy cluster allows both global and regional STS endpoint, but disconnected cluster only allow regional STS endpoint, some operators always access global endpoint (also in disconnected cluster) that caused the problem.

If that, can we say for disconnected + STS cluster, setting custom endpoints is a MUST.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/provision/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-provision-chain.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    - ref: ipi-conf-aws
+    - ref: ipi-conf-aws-custom-endpoints
+    - ref: ipi-conf-mirror
+    - ref: ipi-install-monitoringpvc
+    - ref: proxy-config-generate
+    - ref: ipi-install-install-aws
+    - ref: aws-provision-tags-for-byo-vpc
+    - ref: enable-qe-catalogsource-disconnected
+    - ref: set-sample-operator-disconnected
+    - ref: ipi-install-times-collection
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/provision/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-provision-chain.yaml:21]: I am trying to tidy up the workflows in https://github.com/openshift/release/pull/33289, suggest to move `ref: ipi-install-times-collection` behind `ref: ipi-install-install-aws` to keep consistent.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.metadata.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+	"path": "cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.yaml",
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.metadata.json:2]: To keep consistent with the existing workflow naming, suggest to rename `sts` to `cco-manual-security-token-service`

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/deprovision/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-deprovision-chain.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+chain:
+  as: cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-deprovision
+  steps:
+  - ref: gather-aws-console
+  - chain: gather
+  - ref: gather-proxy
+  - ref: ipi-deprovision-deprovision
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/deprovision/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-deprovision-chain.yaml:7]: I am trying to tidy up deprovision chains in https://github.com/openshift/release/pull/33289, so suggest to wrapper the above lines with `chain: cucushift-installer-rehearse-aws-ipi-deprovision`

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/deprovision/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-deprovision-chain.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+chain:
+  as: cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-deprovision
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/deprovision/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-deprovision-chain.yaml:2]: If "custom-endpoints" is a must for sts + disconnected clusters, maybe we can delete `custom-endpoints` subfolder from the path?

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+workflow:
+  as: cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.yaml:2]: If "custom-endpoints" is a must for sts + disconnected clusters, maybe we can delete custom-endpoints subfolder from the path?

File: ci-operator/step-registry/ipi/conf/aws/custom-endpoints/ipi-conf-aws-custom-endpoints-commands.sh (hunk 1 above)
Comment (yunjiang29) [reply ci-operator/step-registry/ipi/conf/aws/custom-endpoints/ipi-conf-aws-custom-endpoints-commands.sh:11]: `"DEFAULT_ENDPOINT"` means:
1. `serviceEndpoints.[serviceName]` field will be set in install-config.yaml
2. The endpoint value is  `https://${service_name}.${REGION}.amazonaws.com`

`""` means:
1. `serviceEndpoints.[serviceName]` field will NOT be set in install-config.yaml

We don't want to set EP in `install-config.yaml` if it is not added explicitly. 
For the following two configurations, some operators may have totally different behaviors (e.g. [OCPBUGS-1574](https://issues.redhat.com/browse/OCPBUGS-1574)), and we'd like to leave this control to the user.
* a) setting EP to the same as the default one in AWS SDK (`https://${service_name}.${REGION}.amazonaws.com`) explicitly in `install-config.yaml`
* b) don't set RP in config file.

**Comment by yunjiang29** [comment]
> > Proxy cluster allows both global and regional STS endpoint, but disconnected cluster only allow regional STS endpoint, some operators always access global endpoint (also in disconnected cluster) that caused the problem.
> 
> If that, can we say for disconnected + STS cluster, setting custom endpoints is a MUST.

Per my latest test, custom ep is not necessary for disconnect-sts cluster.
But we still need this capability to cover some customer cases.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.yaml (hunk 6 above)
Comment (yunjiang29) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.yaml:2]: Per the previous discussions, we want to add two workflows:
1. Pure disconnect-sts cluster to cover general configuration.
2. Disconnect-sts cluster + some specific EPs to cover customer cases.

Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/sts/custom-endpoints/cucushift-installer-rehearse-aws-ipi-disconnected-private-sts-custom-endpoints-workflow.yaml:2]: get it.

File: ci-operator/step-registry/ipi/conf/aws/custom-endpoints/ipi-conf-aws-custom-endpoints-commands.sh (hunk 1 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/custom-endpoints/ipi-conf-aws-custom-endpoints-commands.sh:11]: > "" means: serviceEndpoints.[serviceName] field will NOT be set in install-config.yaml

Sounds reasonable.

**Comment by yunjiang29** [comment]
Deploying disconnect-STS cluster is blocked by:
* 4.12 https://issues.redhat.com/browse/OCPBUGS-2882
* 4.11 https://issues.redhat.com/browse/OCPBUGS-2883
* 4.10 https://issues.redhat.com/browse/OCPBUGS-2884

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/cco-manual-security-token-service/deprovision/cucushift-installer-rehearse-aws-ipi-disconnected-private-cco-manual-security-token-service-deprovision-chain.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+chain:
+  as: cucushift-installer-rehearse-aws-ipi-disconnected-private-cco-manual-security-token-service-deprovision
+  steps:
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/cco-manual-security-token-service/deprovision/cucushift-installer-rehearse-aws-ipi-disconnected-private-cco-manual-security-token-service-deprovision-chain.yaml:3]: Suggest to insert `gather-mirror-registry` step here.

Comment (yunjiang29) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/disconnected/private/cco-manual-security-token-service/deprovision/cucushift-installer-rehearse-aws-ipi-disconnected-private-cco-manual-security-token-service-deprovision-chain.yaml:3]: Done

**Comment by yunjiang29** [comment]
@jianlinliu it's ready for merge.

**Pull request merged**